### PR TITLE
Implement PKCS authentication

### PR DESCRIPTION
**Pull request by JavaJoeS**
To enable a PKCS12 context, use org.eclipse.core.pki.util.TemplateForPKIfile. Run it and it creates a template in your .eclipse directory.  Edit as needed and then run eclipse.  You should end up with a .pki file and your pki certificate passwd will get
encrypted and be available for subsequent eclipse instances.   The context for SSL will be available for any eclipse bundle
that checks to see if it already exists prior to creating one, which is how it should work.

See this issue for details
- https://github.com/eclipse-platform/eclipse.platform/issues/728

### CODE REVIEW

File: team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/FingerprintX509.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	private static final char[] HEX= {'0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'};
+
+	// private static final String ALG="MD5"; //$NON-NLS-1$
+	public String getFingerPrint(Certificate cert, String alg) {
+		String fingerPrint=null;
+		byte[] encodedCert=null;
+
+		try {
+			encodedCert = cert.getEncoded();
+			MessageDigest md = MessageDigest.getInstance(alg);
```
[the remark this answers is not in the document]
Comment (JavaJoeS) [reply team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/FingerprintX509.java:32]: Updating to SHA-256

File: team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/util/NormalizeAES256.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+			byte[] iv = new byte[16];
+			System.arraycopy(encryptedData, 0, iv, 0, iv.length);
+			IvParameterSpec ivspec = new IvParameterSpec(iv);
+
+			SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256"); //$NON-NLS-1$
+			KeySpec spec = new PBEKeySpec(secretKey.toCharArray(), salt.getBytes(), ITERATION_COUNT, KEY_LENGTH);
+			SecretKey tmp = factory.generateSecret(spec);
+			SecretKeySpec secretKeySpec = new SecretKeySpec(tmp.getEncoded(), "AES"); //$NON-NLS-1$
+
+			Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding"); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (JavaJoeS) [reply team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/util/NormalizeAES256.java:31]: Replaced with GCM Cipher

File: team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/util/SecureAES256.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	        byte[] iv = new byte[16];
+	        secureRandom.nextBytes(iv);
+	        IvParameterSpec ivspec = new IvParameterSpec(iv);
+
+			SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256"); //$NON-NLS-1$
+	        KeySpec spec = new PBEKeySpec(secretKey.toCharArray(), salt.getBytes(), ITERATION_COUNT, KEY_LENGTH);
+	        SecretKey tmp = factory.generateSecret(spec);
+			SecretKeySpec secretKeySpec = new SecretKeySpec(tmp.getEncoded(), "AES"); //$NON-NLS-1$
+
+			Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding"); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (JavaJoeS) [reply team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/util/SecureAES256.java:32]: Replaced with GCM Cipher

File: team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/AuthenticationBase.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		TrustManager[] trustManagers = new TrustManager[1];
+		try {
+			DebugLogger.printDebug("In setSSLContext initialize TLS"); //$NON-NLS-1$
+			// sslContext = SSLContext.getInstance("TLS");
+			sslContext = SSLContext.getInstance("TLSv1.3"); //$NON-NLS-1$
+			CustomKeyManager manager = new CustomKeyManager(keyStore, "".toCharArray(), null); //$NON-NLS-1$
+			manager.setSelectedFingerprint(getFingerprint());
+			keyManagers[0] = manager;
+			trustManagers[0] = new CustomTrustManager(keyStore);
+			sslContext.init(keyManagers, trustManagers, new SecureRandom());
```
[the remark this answers is not in the document]
Comment (JavaJoeS) [reply team/bundles/org.eclipse.core.pki/src/org/eclipse/core/pki/AuthenticationBase.java:196]: Set CustomTrustManager to not accept just any.

**Comment by merks** [comment]
I see you've been asking for reviews...

This PR is not linked to the following issue so I had to hunt for this related issue:

https://github.com/eclipse-platform/eclipse.platform/issues/728

Also, you've typically been making chains of commits which makes it hard to review and it means this needs to be squashed in the end, which is even more work.  Generally it would be better to have one commit, amend it each time, and force push it.

The title of the PR is meaningless.

In the end, I can find no overview or description of what this is all about.  Why would I use it?  For what would I use it?

The PR contains unrelated changes:

image

The code itself contains many TODOs, lots of e.printStackTrace(), System.err.println, and System.out.println.  It seems all exceptions ignored, other than print statement.  I've not see DebugLogger.printDebug before.  What's that?

Things like this don't follow Java naming conventions:
```
	private static final String[] cdir = { "C:\\Progra~1\\Java", "C:\\Progra~2\\Java" }; //$NON-NLS-1$ //$NON-NLS-2$
```
I would hope that nothing needs to find Java in some specific location/drive on my machine...

There's effectively no documentation here.  There are no tests either.

To me this does not look finished and does not look ready for review.   When it is ready, there needs to be documentation to describe the purpose and the impact.  There need to be tests.  In the end, someone with expertise in this area will be needed.  It looks like days of work would be needed to review this properly making the motivation even more significant.

**Review by JavaJoeS — COMMENTED** [review]
Many of these files have nothing to do with my submitted bundles and I dont understand why they are appearing as changed, because I did not change them, or rather knowingly change them.

**Comment by laeubi** [comment]
>  because I did not change them, or rather knowingly change them.

@JavaJoeS in your IDE select any file you don't wanted to have changed, then select Replace With > Previous Revision

after that, amend your commit and force push it, these should then be gone, before you add something to the index make sure you really want to submit that change or revert it otherwise.

**Comment by laeubi** [comment]
> https://github.com/eclipse-platform/eclipse.platform/commit/d0989e3655c30e96b0e03354df85e1a307744c33 Merge branch 'eclipse-platform:master' into master

@JavaJoeS better _squash_ all your commits into one and _rebase_ instead of _merge_ master branch

**Comment by merks** [comment]
Yes, and use amend followed by a force push to keep the branch to a single commit.

**Comment by JavaJoeS** [comment]
Attempting to squash down and merge

**Pull request closed without merging**